x/mongo/driver/topology/server_test.go

```diff
@@ -54,6 +54,217 @@ func (cncd *channelNetConnDialer) DialContext(_ context.Context, _, _ string) (n
+					// error returned by DialContext() to be treated as a timeout caused by reaching
+					// connectTimeoutMS.
+					ctx, cancel := context.WithDeadline(ctx, time.Now().Add(-1*time.Hour))
+					defer cancel()
+					return d.DialContext(ctx, network, addr)
+				})
+			},
+			expectErr:         true,
+			expectPoolCleared: true,
+		},
+		{
+			desc: "Parent context deadline exceeded error during handshake should not clear the pool",
+			ctxFn: func() (context.Context, context.CancelFunc) {
+				return context.WithTimeout(context.Background(), 100*time.Millisecond)
+			},
+			handshaker: func(Handshaker) Handshaker {
+				h := auth.Handshaker(nil, &auth.HandshakeOptions{})
+				return &testHandshaker{
+					getHandshakeInformation: func(ctx context.Context, addr address.Address, c driver.Connection) (driver.HandshakeInformation, error) {
+						// Sleep for at least 150ms and expect the context passed to
+						// server.Connection() to time out during the sleep. Expect the error
+						// returned by GetHandshakeInformation() to be treated as a timeout caused
+						// by an operation-scoped deadline.
+						// E.g. FindOne(context.WithTimeout(...))
+						time.Sleep(150 * time.Millisecond)
+						return h.GetHandshakeInformation(ctx, addr, c)
+					},
+				}
+			},
+			expectErr:         true,
+			expectPoolCleared: false,
+		},
+		{
+			desc: "Child context deadline exceeded error during handshake should clear the pool",
+			ctxFn: func() (context.Context, context.CancelFunc) {
+				return context.WithTimeout(context.Background(), 100*time.Millisecond)
+			},
+			handshaker: func(Handshaker) Handshaker {
+				h := auth.Handshaker(nil, &auth.HandshakeOptions{})
+				return &testHandshaker{
+					getHandshakeInformation: func(ctx context.Context, addr address.Address, c driver.Connection) (driver.HandshakeInformation, error) {
+						// Wrap the context in a context with an already-exceeded deadline. Expect
+						// the error returned by GetHandshakeInformation() to be treated as a
+						// timeout caused by connectTimeoutMS.
+						ctx, cancel := context.WithDeadline(ctx, time.Now().Add(-1*time.Hour))
+						defer cancel()
+						return h.GetHandshakeInformation(ctx, addr, c)
+					},
+				}
+			},
+			expectErr:         true,
+			expectPoolCleared: true,
+		},
+		{
+			desc: "Dial errors unrelated to context timeouts should clear the pool",
+			ctxFn: func() (context.Context, context.CancelFunc) {
+				return context.WithTimeout(context.Background(), 1*time.Second)
+			},
+			dialer: func(Dialer) Dialer {
+				var d net.Dialer
+				return DialerFunc(func(ctx context.Context, _, _ string) (net.Conn, error) {
+					// Try to dial an invalid TCP address and expect an error.
+					return d.DialContext(ctx, "tcp", "300.0.0.0:nope")
+				})
+			},
+			expectErr:         true,
+			expectPoolCleared: true,
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(tc.desc, func(t *testing.T) {
+			t.Parallel()
```

Comment:
Oh nice! I assume the `tc := tc` above is necessary to capture the value of `tc` in each of the parallel goroutines.

Reply:
That's correct, otherwise the test `func`s will not capture the actual range variable and will just use whatever the pointer value when they're actually run (typically the last value in the slice). Not doing that can cause some nasty bugs, like only the last test case ever being run.